File: benchmarks/compare.zig
//! Volt vs Go comparative benchmark runner.
//!
//! Builds both implementations in release mode, runs each, parses
//! their JSON output, and prints a side-by-side comparison table.
//!
//! Usage:
//!   zig build compare
//!
//! Both sides run the SAME workloads with the SAME iteration counts,
//! median over 11 samples. We don't run them concurrently — sequential
//! execution avoids cache/thermal interference between them.

const std = @import("std");

const Color = struct {
    const red = "\x1b[0;31m";
    const green = "\x1b[0;32m";
    const yellow = "\x1b[1;33m";
    const cyan = "\x1b[0;36m";
    const bold = "\x1b[1m";
    const dim = "\x1b[2m";
    const reset = "\x1b[0m";
};

const Results = struct {
    yield_one_way_ns: u64 = 0,
    spawn_join_ns: u64 = 0,
    spawn_waitgroup_ns: u64 = 0,
    channel_spsc_16_ns: u64 = 0,
    mutex_contended_8_ns: u64 = 0,

    pub fn parseFrom(json: []const u8, allocator: std.mem.Allocator) !Results {
        const parsed = try std.json.parseFromSlice(Results, allocator, json, .{});
        defer parsed.deinit();
        return parsed.value;
    }
};

const Bench = struct {
    name: []const u8,
    field: []const u8,
    unit: []const u8 = "ns/op",
};

const BENCHES = [_]Bench{
    .{ .name = "yield (one-way ctx switch)", .field = "yield_one_way_ns" },
    .{ .name = "spawn + waitgroup-wait", .field = "spawn_waitgroup_ns" },
    .{ .name = "spawn + per-coro join", .field = "spawn_join_ns" },
    .{ .name = "channel SPSC cap=16", .field = "channel_spsc_16_ns" },
    .{ .name = "mutex contended (8 coros)", .field = "mutex_contended_8_ns" },
};

fn runCommand(
    gpa: std.mem.Allocator,
    io: std.Io,
    argv: []const []const u8,
    cwd: ?[]const u8,
) !struct { stdout: []u8, stderr: []u8, success: bool } {
    const result = try std.process.run(gpa, io, .{
        .argv = argv,
        .cwd = if (cwd) |dir| .{ .path = dir } else .inherit,
    });
    const success = switch (result.term) {
        .exited => |code| code == 0,
        else => false,
    };
    return .{ .stdout = result.stdout, .stderr = result.stderr, .success = success };
}

fn fieldValue(r: Results, field: []const u8) u64 {
    inline for (@typeInfo(Results).@"struct".fields) |f| {
        if (std.mem.eql(u8, f.name, field)) {
            return @field(r, f.name);
        }
    }
    unreachable;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;

    std.debug.print("{s}{s}Building Go benchmark...{s}\n", .{ Color.bold, Color.cyan, Color.reset });
    {
        const r = try runCommand(
            allocator,
            io,
            &.{ "go", "build", "-o", "compare-bin", "compare.go" },
            "benchmarks/go",
        );
        defer allocator.free(r.stdout);
        defer allocator.free(r.stderr);
        if (!r.success) {
            std.debug.print("{s}Go build failed{s}\n{s}\n", .{ Color.red, Color.reset, r.stderr });
            return error.GoBuildFailed;
        }
    }

    std.debug.print("{s}{s}Running Volt benchmark (this takes ~30-60s)...{s}\n", .{ Color.bold, Color.yellow, Color.reset });
    const volt_r = try runCommand(
        allocator,
        io,
        &.{"./zig-out/bin/volt-bench-compare"},
        null,
    );
    defer allocator.free(volt_r.stdout);
    defer allocator.free(volt_r.stderr);
    if (!volt_r.success) {
        std.debug.print("{s}Volt bench failed{s}\n", .{ Color.red, Color.reset });
        return error.VoltBenchFailed;
    }
    // Volt emits JSON via std.debug.print (→ stderr).
    const volt_res = Results.parseFrom(volt_r.stderr, allocator) catch |err| {
        std.debug.print("{s}Failed to parse Volt JSON{s}: {any}\nstderr was:\n{s}\n", .{ Color.red, Color.reset, err, volt_r.stderr });
        return err;
    };

    std.debug.print("{s}{s}Running Go benchmark (this takes ~30-60s)...{s}\n", .{ Color.bold, Color.yellow, Color.reset });
    const go_r = try runCommand(
        allocator,
        io,
        &.{"./compare-bin"},
        "benchmarks/go",
    );
    defer allocator.free(go_r.stdout);
    defer allocator.free(go_r.stderr);
    if (!go_r.success) {
        std.debug.print("{s}Go bench failed{s}\n", .{ Color.red, Color.reset });
        return error.GoBenchFailed;
    }
    const go_res = Results.parseFrom(go_r.stdout, allocator) catch |err| {
        std.debug.print("{s}Failed to parse Go JSON{s}: {any}\nstdout was:\n{s}\n", .{ Color.red, Color.reset, err, go_r.stdout });
        return err;
    };

    std.debug.print("\n", .{});
    std.debug.print("{s}╔════════════════════════════════════════════════════════════════════╗{s}\n", .{ Color.bold, Color.reset });
    std.debug.print("{s}║          Volt vs Go — comparative benchmarks                       ║{s}\n", .{ Color.bold, Color.reset });
    std.debug.print("{s}╚════════════════════════════════════════════════════════════════════╝{s}\n", .{ Color.bold, Color.reset });
    std.debug.print("\n", .{});

    std.debug.print("  {s}{s:<34} {s:>14} {s:>14} {s:>10}{s}\n", .{
        Color.bold,                            "Workload",
        "Volt",                                "Go",
        "Volt / Go",
        Color.reset,
    });
    std.debug.print("  {s}{s}{s}\n", .{ Color.dim, "─" ** 78, Color.reset });

    for (BENCHES) |b| {
        const volt_ns = fieldValue(volt_res, b.field);
        const go_ns = fieldValue(go_res, b.field);
        const ratio: f64 = @as(f64, @floatFromInt(volt_ns)) / @as(f64, @floatFromInt(go_ns));

        const ratio_color = if (ratio < 0.95) Color.green else if (ratio < 1.05) Color.yellow else Color.red;
        const ratio_glyph: []const u8 = if (ratio < 0.5)
            " ★★"
        else if (ratio < 0.95)
            " ★ "
        else if (ratio < 1.05)
            " ≈ "
        else
            "   ";

        std.debug.print("  {s:<34} {d:>10} {s:<3} {d:>10} {s:<3} {s}{d:>6.2}×{s}{s}\n", .{
            b.name,
            volt_ns,
            b.unit,
            go_ns,
            b.unit,
            ratio_color,
            ratio,
            Color.reset,
            ratio_glyph,
        });
    }

    std.debug.print("\n", .{});
    std.debug.print("  {s}★★ = ≥2× faster   ★ = 5-50% faster   ≈ = within 5%   (blank) = slower{s}\n", .{ Color.dim, Color.reset });
    std.debug.print("\n", .{});
}

File: benchmarks/go/compare.go
// Volt vs Go comparative benchmarks (Go side).
//
// Emits JSON with median over N iterations matching the Volt-side bench.
// Same workloads, same iteration counts, same GOMAXPROCS as Volt's
// default worker count (= NumCPU).
//
// Build/run: see ../compare.zig orchestrator.

package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"runtime"
	"sort"
	"sync"
	"time"
)

const (
	WarmupIters = 3
	BenchIters  = 11
)

// ─────────────────────────────────────────────────────────────────────
// Workload: yield ping-pong (one-way via runtime.Gosched)
// ─────────────────────────────────────────────────────────────────────
//
// Volt's "yield ping-pong" measures ctx_swap cost between two coros.
// Go's analog: a single goroutine calling runtime.Gosched() in a loop.
// Each Gosched is conceptually one round-trip through the scheduler.
// We divide wall time by (iters * 2) to match Volt's "one-way" report.

func benchYield(iters int) uint64 {
	done := make(chan struct{})
	var t0 time.Time
	go func() {
		t0 = time.Now()
		for i := 0; i < iters; i++ {
			runtime.Gosched()
		}
		close(done)
	}()
	<-done
	wall := time.Since(t0).Nanoseconds()
	return uint64(wall) / uint64(iters*2)
}

// ─────────────────────────────────────────────────────────────────────
// Workload: spawn + join (newproc + wait via channel)
// ─────────────────────────────────────────────────────────────────────

func benchSpawnJoin(n int) uint64 {
	t0 := time.Now()
	var wg sync.WaitGroup
	wg.Add(n)
	for i := 0; i < n; i++ {
		go func() { wg.Done() }()
	}
	wg.Wait()
	wall := time.Since(t0).Nanoseconds()
	return uint64(wall) / uint64(n)
}

// ─────────────────────────────────────────────────────────────────────
// Workload: channel SPSC cap=16
// ─────────────────────────────────────────────────────────────────────

func benchChannelSpsc(n int) uint64 {
	ch := make(chan uint64, 16)
	done := make(chan struct{})

	t0 := time.Now()
	go func() {
		for i := 0; i < n; i++ {
			ch <- uint64(i)
		}
		close(ch)
	}()
	go func() {
		var sum uint64 = 0
		for v := range ch {
			sum += v
		}
		_ = sum
		close(done)
	}()
	<-done
	wall := time.Since(t0).Nanoseconds()
	return uint64(wall) / uint64(n)
}

// ─────────────────────────────────────────────────────────────────────
// Workload: TCP echo, N concurrent clients, K requests each.
// Match Volt's params: msg=1024 bytes, reqs_per_client=16.
// ─────────────────────────────────────────────────────────────────────

const (
	tcpMsgSize        = 1024
	tcpReqsPerClient  = 16
)

func tcpEchoConn(c net.Conn) {
	defer c.Close()
	buf := make([]byte, tcpMsgSize)
	for i := 0; i < tcpReqsPerClient; i++ {
		if _, err := io.ReadFull(c, buf); err != nil {
			return
		}
		if _, err := c.Write(buf); err != nil {
			return
		}
	}
}

func tcpClient(addr string, wg *sync.WaitGroup) {
	defer wg.Done()
	c, err := net.Dial("tcp", addr)
	if err != nil {
		return
	}
	defer c.Close()
	send := make([]byte, tcpMsgSize)
	for i := range send {
		send[i] = byte(i & 0xff)
	}
	recv := make([]byte, tcpMsgSize)
	for i := 0; i < tcpReqsPerClient; i++ {
		if _, err := c.Write(send); err != nil {
			return
		}
		if _, err := io.ReadFull(c, recv); err != nil {
			return
		}
	}
}

func benchTcpEcho(nClients int) uint64 {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		fmt.Fprintf(os.Stderr, "listen: %v\n", err)
		os.Exit(1)
	}
	defer ln.Close()
	addr := ln.Addr().String()

	// Server accept loop, exactly nClients connections.
	serverDone := make(chan struct{})
	go func() {
		defer close(serverDone)
		for i := 0; i < nClients; i++ {
			c, err := ln.Accept()
			if err != nil {
				return
			}
			go tcpEchoConn(c)
		}
	}()

	var wg sync.WaitGroup
	wg.Add(nClients)
	t0 := time.Now()
	for i := 0; i < nClients; i++ {
		go tcpClient(addr, &wg)
	}
	wg.Wait()
	wall := time.Since(t0).Nanoseconds()
	<-serverDone

	totalRtts := uint64(nClients) * uint64(tcpReqsPerClient)
	return uint64(wall) / totalRtts
}

// ─────────────────────────────────────────────────────────────────────
// Workload: channel pipeline (producer → middle → consumer).
// Two channels (cap 64), modest per-message work.
// ─────────────────────────────────────────────────────────────────────

func benchChannelPipeline(n int) uint64 {
	in := make(chan uint64, 64)
	out := make(chan uint64, 64)
	done := make(chan struct{})

	t0 := time.Now()
	go func() {
		for i := 0; i < n; i++ {
			in <- uint64(i)
		}
		close(in)
	}()
	go func() {
		for v := range in {
			h := v
			h ^= h >> 13
			h *= 0x100000001b3
			h ^= h >> 23
			out <- h
		}
		close(out)
	}()
	go func() {
		var sum uint64
		for v := range out {
			sum += v
		}
		_ = sum
		close(done)
	}()
	<-done
	wall := time.Since(t0).Nanoseconds()
	return uint64(wall) / uint64(n)
}

// ─────────────────────────────────────────────────────────────────────
// Workload: mutex contended (8 × 50k)
// ─────────────────────────────────────────────────────────────────────

func benchMutex(workers int, iters int) uint64 {
	var mu sync.Mutex
	var counter uint64

	var wg sync.WaitGroup
	wg.Add(workers)

	t0 := time.Now()
	for w := 0; w < workers; w++ {
		go func() {
			for i := 0; i < iters; i++ {
				mu.Lock()
				counter++
				mu.Unlock()
			}
			wg.Done()
		}()
	}
	wg.Wait()
	wall := time.Since(t0).Nanoseconds()

	if counter != uint64(workers*iters) {
		fmt.Fprintf(os.Stderr, "mutex bench correctness: expected %d, got %d\n",
			workers*iters, counter)
		os.Exit(1)
	}
	return uint64(wall) / uint64(workers*iters)
}

// ─────────────────────────────────────────────────────────────────────
// Runner — median over N iterations
// ─────────────────────────────────────────────────────────────────────

func runMedian(fn func() uint64) uint64 {
	for i := 0; i < WarmupIters; i++ {
		_ = fn()
	}
	samples := make([]uint64, BenchIters)
	for i := range samples {
		samples[i] = fn()
	}
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	return samples[BenchIters/2]
}

type Results struct {
	YieldOneWayNs     uint64 `json:"yield_one_way_ns"`
	SpawnJoinNs       uint64 `json:"spawn_join_ns"`
	SpawnWaitgroupNs  uint64 `json:"spawn_waitgroup_ns"`
	ChannelSpsc16Ns   uint64 `json:"channel_spsc_16_ns"`
	MutexContended8Ns uint64 `json:"mutex_contended_8_ns"`
	TcpEchoRttNs      uint64 `json:"tcp_echo_rtt_ns"`
	ChannelPipelineNs uint64 `json:"channel_pipeline_ns"`
}

func main() {
	// Match Volt's default worker count (uses NumCPU). GOMAXPROCS
	// defaults to NumCPU too, so this is implicit on most systems —
	// set it explicitly here for clarity and reproducibility.
	runtime.GOMAXPROCS(runtime.NumCPU())

	r := Results{
		YieldOneWayNs: runMedian(func() uint64 { return benchYield(100_000) }),
		// spawn_join: per-coroutine join (matches Volt's `for j |.| j.join()`).
		// Go's WaitGroup pattern is more idiomatic — see spawn_waitgroup.
		SpawnJoinNs: runMedian(func() uint64 { return benchSpawnJoin(10_000) }),
		// spawn_waitgroup: spawn + single batched wait. Volt's analog
		// uses a Notify+atomic counter; both wait for ALL coros at once.
		SpawnWaitgroupNs:  runMedian(func() uint64 { return benchSpawnJoin(10_000) }),
		ChannelSpsc16Ns:   runMedian(func() uint64 { return benchChannelSpsc(100_000) }),
		MutexContended8Ns: runMedian(func() uint64 { return benchMutex(8, 50_000) }),
		TcpEchoRttNs:      runMedian(func() uint64 { return benchTcpEcho(64) }),
		ChannelPipelineNs: runMedian(func() uint64 { return benchChannelPipeline(50_000) }),
	}

	out, err := json.MarshalIndent(r, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "json marshal: %v\n", err)
		os.Exit(1)
	}
	fmt.Println(string(out))
}
